tempesta_fw/avx_routines.h

```diff
@@ -0,0 +1,152 @@
+        c3 = _mm256_and_si256(c1, c3); \
+        c3 = _mm256_andnot_si256(c2, c3); \
+        out = _mm256_or_si256(in, c3); \
+    } while(0);
+
+#define AVX_MATCH_HEADER(result, text, charset) do { \
+        __m256i mask1 = _mm256_shuffle_epi8( \
+                    charset, text); \
+        __m256i mask2 = _mm256_and_si256( \
+                    _mm256_shuffle_epi32(__avx_constants2, __AVX_C2_CHARSET_M1), \
+                    _mm256_srli_epi16(text, 4)); \
+        __m256i mask3 = _mm256_shuffle_epi8( \
+                    _mm256_shuffle_epi32(__avx_constants2, __AVX_C2_CHARSET_M2), \
+                    mask2); \
+        __m256i mask4 = _mm256_and_si256( \
+                    mask3, mask1); \
+        result = _mm256_cmpeq_epi8( \
+                    mask4, _mm256_setzero_si256()); \
+    } while(0);
+
+//==================================================
+//DATA#  0low 0hi  1low 1hi  2low 2hi  3low 3hi
+//DWORD#
+// 0     Host Tran X-Fo Cont Cach Cook Conn User
+// 1          sfer rwar ent- e-Co ie   ecti -age
+// 2     Type -Enc ded- Leng ntro g    on   nt
+// 3          odin For  th   l
+//==================================================
+
+#define AVX_MATCH_HEADER_NAME(result, text) do {\
+    /*mask 1 = rows 0-1, columns 0low-1hi*/\
+    __m256i mask1 = _mm256_cmpeq_epi64(\
+        _mm256_and_si256(_mm256_permute4x64_epi64(text, 0x00),\
+                         _mm256_load(__c_avx_header_names)),\
+        _mm256_load((__c_avx_header_names+32)));\
+    /*mask 2 = rows 0-1, columns 2low-3hi*/\
+    __m256i mask2 = _mm256_cmpeq_epi64(\
+        _mm256_and_si256(_mm256_permute4x64_epi64(text, 0x00),\
+                         _mm256_load(__c_avx_header_names+64)),\
+        _mm256_load((__c_avx_header_names+96)));\
+    /*mask 3 = rows 2-3, columns 0low-1hi*/\
+    __m256i mask3 = _mm256_cmpeq_epi64(\
+        _mm256_and_si256(_mm256_permute4x64_epi64(text, 0x55),\
+                         _mm256_load(__c_avx_header_names+128)),\
+        _mm256_load((__c_avx_header_names+160)));\
+    /*mask 4 = rows 2-3, columns 2low-3hi*/\
+    __m256i mask4 = _mm256_cmpeq_epi64(\
+        _mm256_and_si256(_mm256_permute4x64_epi64(text, 0x95),\
+                         _mm256_load(__c_avx_header_names+192)),\
+        _mm256_load((__c_avx_header_names+224)));\
+    /*extract mask for host*/\
+    int bm = 0x1 & _mm256_movemask_epi8(mask1);\
+    /*combine mask1 and mask3*/\
+    mask1 = _mm256_and_si256(_mm256_permute4x64_epi64(mask1, 0xe7),\
+                             mask3);\
+    mask1 = _mm256_and_si256(mask1, _mm256_blend_epi32(mask4, 0x0c);\
+    /*combine mask2 and mask4*/\
+    mask2 = _mm256_and_si256(mask2, _mm256_blend_epi32(mask4, 0xf3);\
+    bm |= 0x05040302 & _mm256_movemask_epi8(mask1);\
+    bm |= 0x09080706 & _mm256_movemask_epi8(mask2);\
+    bm = bm | (bm>>16);\
+    bm = bm | (bm>>8);\
+    result = 0xFF & bm;} while(0);
+
+#define AVX_STRID_1(n) do {\
+    static const int ids[9] = {\
```

Comment:
Fixed in latest commit

2016-06-27 13:01 GMT+04:00 Ilya V. Matveychikov <notifications@github.com>:

> In tempesta_fw/avx_routines.h
> <https://github.com/tempesta-tech/tempesta/pull/573#discussion_r68543362>:
>
> > +    /*extract mask for host*/\
> > +    int bm = 0x1 & _mm256_movemask_epi8(mask1);\
> > +    /*combine mask1 and mask3*/\
> > +    mask1 = _mm256_and_si256(_mm256_permute4x64_epi64(mask1, 0xe7),\
> > +                             mask3);\
> > +    mask1 = _mm256_and_si256(mask1, _mm256_blend_epi32(mask4, 0x0c);\
> > +    /*combine mask2 and mask4*/\
> > +    mask2 = _mm256_and_si256(mask2, _mm256_blend_epi32(mask4, 0xf3);\
> > +    bm |= 0x05040302 & _mm256_movemask_epi8(mask1);\
> > +    bm |= 0x09080706 & _mm256_movemask_epi8(mask2);\
> > +    bm = bm | (bm>>16);\
> > +    bm = bm | (bm>>8);\
> > +    result = 0xFF & bm;} while(0);
> > +
> > +#define AVX_STRID_1(n) do {\
> > +    static const int ids[9] = {\
>
> There must be ids[10] here.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tempesta-tech/tempesta/pull/573/files/a61f733b2e57753cf48c99f68ef520ef56d4798b#r68543362>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABnrSY8IpX58vPav3lyOzZ1wqNQbqbHLks5qP5F4gaJpZM4I9yWw>
> .
>